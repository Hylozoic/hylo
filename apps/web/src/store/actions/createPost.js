import { get, uniqueId } from 'lodash/fp'
import { TextHelpers, AnalyticsEvents } from '@hylo/shared'
import createPostMutation from '@graphql/mutations/createPostMutation'
import { CREATE_POST } from 'store/constants'

export default function createPost (postParams) {
  const query = createPostMutation

  const {
    acceptContributions,
    completionAction,
    completionActionSettings,
    details,
    donationsLink,
    endTime,
    eventInviteeIds = [],
    fileUrls,
    fundingRoundId,
    groups,
    imageUrls,
    linkPreview,
    linkPreviewFeatured,
    localId,
    location,
    locationId,
    memberIds = [],
    isAnonymousVote,
    isPublic,
    isStrictProposal,
    projectManagementLink,
    proposalOptions,
    votingMethod,
    quorum,
    sendAnnouncement,
    startTime,
    timezone,
    title,
    topicNames,
    trackId,
    type
  } = postParams
  const linkPreviewId = linkPreview && linkPreview.id
  const groupIds = groups.map(c => c.id)

  return {
    type: CREATE_POST,
    graphql: {
      query,
      variables: {
        acceptContributions,
        announcement: sendAnnouncement,
        completionAction,
        completionActionSettings,
        details,
        donationsLink,
        endTime: endTime && endTime.valueOf(),
        eventInviteeIds,
        fileUrls,
        fundingRoundId,
        groupIds,
        imageUrls,
        isAnonymousVote,
        isPublic,
        isStrictProposal,
        linkPreviewId,
        linkPreviewFeatured,
        localId: localId || uniqueId('post_'), // to match the optimistically created post and replace it with the real one
        location,
        locationId,
        memberIds,
        projectManagementLink,
        proposalOptions,
        votingMethod,
        quorum,
        startTime: startTime && startTime.valueOf(),
        timezone,
        title,
        topicNames,
        trackId,
        type
      }
    },
    meta: {
      extractModel: {
        modelName: 'Post',
        getRoot: get('createPost')
      },
      analytics: {
        eventName: AnalyticsEvents.POST_CREATED,
        detailsLength: TextHelpers.textLengthHTML(details),
        fundingRoundId,
        groupId: groupIds,
        isAnnouncement: sendAnnouncement,
        isPublic,
        topics: topicNames,
        trackId,
        type
      },
      type,
      fundingRoundId,
      groupIds,
      trackId
    }
  }
}
