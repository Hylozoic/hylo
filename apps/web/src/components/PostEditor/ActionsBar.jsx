import { MapPin, SendHorizontal } from 'lucide-react'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch } from 'react-redux'
import Button from 'components/Button'
import Icon from 'components/Icon'
import Tooltip from 'components/Tooltip'
import SendAnnouncementModal from 'components/SendAnnouncementModal'
import UploadAttachmentButton from 'components/UploadAttachmentButton'
import { cn } from 'util/index'

import styles from './PostEditor.module.scss'

export default function ActionsBar ({
  id,
  addAttachment,
  announcementSelected,
  canMakeAnnouncement,
  isEditing,
  groupCount,
  groups,
  invalidMessage,
  loading,
  myAdminGroups,
  setAnnouncementSelected,
  setShowLocation,
  doSave, // Pops up announcement modal first if announcement is selected
  save, // Does actual save
  setIsDirty,
  showAnnouncementModal,
  showLocation,
  showFiles,
  showImages,
  submitButtonLabel,
  toggleAnnouncementModal,
  type,
  valid
}) {
  const dispatch = useDispatch()
  const { t } = useTranslation()

  return (
    <div className='w-full flex justify-between'>
      <div className='flex items-center gap-2'>
        <UploadAttachmentButton
          type='post'
          id={id}
          attachmentType='image'
          onSuccess={(attachment) => {
            dispatch(addAttachment('post', id, attachment))
            setIsDirty(true)
          }}
          allowMultiple
          disable={showImages}
        >
          <Icon
            name='AddImage'
            className={cn(styles.actionIcon, { [styles.highlightIcon]: showImages })}
            dataTestId='add-image-icon'
          />
        </UploadAttachmentButton>
        <UploadAttachmentButton
          type='post'
          id={id}
          attachmentType='file'
          onSuccess={(attachment) => {
            dispatch(addAttachment('post', id, attachment))
            setIsDirty(true)
          }}
          allowMultiple
          disable={showFiles}
        >
          <Icon
            name='Paperclip'
            className={cn(styles.actionIcon, { [styles.highlightIcon]: showFiles })}
            dataTestId='add-file-icon'
          />
        </UploadAttachmentButton>
        {type !== 'chat' && !showLocation && (
          <span data-tooltip-content={t('Add Location')} data-tooltip-id='location-tt' onClick={() => setShowLocation(true)}>
            <MapPin className={styles.actionIcon} />
          </span>
        )}
        {canMakeAnnouncement && (
          <span data-tooltip-content={t('Send Announcement')} data-tooltip-id='announcement-tt'>
            <Icon
              dataTestId='announcement-icon'
              name='Announcement'
              onClick={() => {
                setAnnouncementSelected(!announcementSelected)
                setIsDirty(true)
              }}
              className={cn(styles.actionIcon, {
                [styles.highlightIcon]: announcementSelected
              })}
            />
            <Tooltip
              effect='solid'
              delayShow={10}
              id='announcement-tt'
            />
          </span>
        )}
        {showAnnouncementModal && (
          <SendAnnouncementModal
            closeModal={toggleAnnouncementModal}
            save={save}
            groupCount={groupCount}
            myAdminGroups={myAdminGroups}
            groups={groups}
          />
        )}
      </div>

      <div className='flex items-center gap-2'>
        <label className='text-xs italic text-foreground/50'>
          {isEditing
            ? t(navigator.platform.includes('Mac') ? 'Option-Enter to save' : 'Alt-Enter to save')
            : t(navigator.platform.includes('Mac') ? 'Option-Enter to post' : 'Alt-Enter to post')}
        </label>
        <Button
          disabled={!valid || loading}
          onClick={doSave}
          className='border-2 border-foreground/30 bg-foreground/30 px-2 py-1 rounded flex items-center'
          dataTipHtml={!valid ? invalidMessage : ''}
          dataFor='submit-tt'
        >
          <SendHorizontal className={!valid || loading ? 'text-foreground/30' : 'text-highlight'} size={18} style={{ display: 'inline' }} />
        </Button>

        <Tooltip
          delay={10}
          position='bottom'
          id='submit-tt'
        />
      </div>
    </div>
  )
}
