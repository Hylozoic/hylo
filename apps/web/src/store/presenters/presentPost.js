import { DateTimeHelpers } from '@hylo/shared'
import presentTopic from 'store/presenters/presentTopic'
import { localeLocalStorageSync } from 'util/locale'

export default function presentPost (post, groupId) {
  if (!post) return null

  // Raw posts came directly from the API, not processed through the model extractor
  // Used in the chat room
  const rawPost = !post.ref

  try {
    const createdAtHumanDate = DateTimeHelpers.humanDate(post.createdAt)
    const editedAtHumanDate = DateTimeHelpers.humanDate(post.editedAt)

    const finalPost = {
      ...(rawPost ? post : post.ref),
      attachments: (rawPost ? post.attachments || [] : post.attachments.toModelArray()).sort((a, b) => a.position - b.position),
      createdTimestamp: createdAtHumanDate,
      creator: post.creator, // needed to load the creator object
      commenters: (rawPost ? post.commenters?.items || [] : post.commenters.toModelArray()),
      editedTimestamp: post.editedAt ? `Edited ${editedAtHumanDate}` : null,
      eventInvitations: (rawPost ? post.eventInvitations?.items || [] : post.eventInvitations.toModelArray()).map(eventInvitation => {
        return {
          response: eventInvitation.response,
          ...(rawPost ? eventInvitation.person : eventInvitation.person.ref)
        }
      }),
      exactCreatedTimestamp: DateTimeHelpers.toDateTime(post.createdAt, { locale: localeLocalStorageSync() }).toFormat('D t ZZZZ'),
      exactEditedTimestamp: DateTimeHelpers.toDateTime(post.editedAt, { locale: localeLocalStorageSync() }).toFormat('D t ZZZZ'),
      fileAttachments: (rawPost ? post.attachments || [] : post.attachments.toModelArray()).filter(a => a.type === 'file').sort((a, b) => a.position - b.position),
      imageAttachments: (rawPost ? post.attachments || [] : post.attachments.toModelArray()).filter(a => a.type === 'image').sort((a, b) => a.position - b.position),
      groups: (rawPost ? post.groups?.items || [] : post.groups.toModelArray()),
      linkPreview: post.linkPreview, // needed to load the link preview object
      location: post.location, // needed to load the location object
      members: (rawPost ? post.members?.items || [] : post.members.toModelArray()).map(person => {
        return {
          ...(rawPost ? person : person.ref),
          skills: (rawPost ? person.skills?.items || [] : person.skills.toModelArray())
        }
      }),
      proposalOptions: (rawPost ? post.proposalOptions?.items || [] : post.proposalOptions.toModelArray()),
      topics: (rawPost ? post.topics?.items || [] : post.topics.toModelArray()).map(topic => presentTopic(topic, {}))
    }
    return finalPost
  } catch (e) {
    console.log('error', e)
  }
}
