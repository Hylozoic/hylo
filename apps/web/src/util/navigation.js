import { get, isEmpty, isNumber, omitBy } from 'lodash/fp'
import qs from 'query-string'
import { host } from 'config/index'
import { isContextGroupSlug } from '@hylo/presenters/GroupPresenter'
import { ALL_GROUPS_CONTEXT_SLUG, MY_CONTEXT_SLUG, PUBLIC_CONTEXT_SLUG } from '@hylo/shared'

export const HYLO_ID_MATCH = '\\d+'
export const POST_ID_MATCH = HYLO_ID_MATCH
const GROUP_SLUG_MATCH = '[^\\\\]+'
// TODO: do this validation elsewhere?
export const OPTIONAL_POST_MATCH = ':detail(post)?/:postId?/:action(new|edit)?'
export const OPTIONAL_NEW_POST_MATCH = ':detail(post)?/:action(new)?' // TODO: need this?
export const POST_DETAIL_MATCH = 'post/:postId/comments?/:commentId?/*'

export const REQUIRED_EDIT_POST_MATCH = ':detail(post)/:postId/:action(edit)'

export const GROUP_DETAIL_MATCH = 'group/:detailGroupSlug'
export const OPTIONAL_GROUP_MATCH = ':detail(group)?/(:detailGroupSlug)?'

// Fundamental URL paths

export function allGroupsUrl () {
  return '/all'
}

export function publicGroupsUrl () {
  return '/public'
}

export function myHomeUrl () {
  return '/my'
}

export function baseUrl ({
  context,
  customViewId,
  defaultUrl = allGroupsUrl(),
  groupSlug,
  memberId, personId, // TODO: switch to one of these?
  topicName,
  view
}) {
  const safeMemberId = personId || memberId

  if (safeMemberId) {
    return personUrl(safeMemberId, groupSlug)
  } else if (view === 'chat' && topicName) {
    return chatUrl(topicName, { context, groupSlug })
  } else if (topicName) {
    return topicUrl(topicName, { context, groupSlug })
  } else if (view) {
    return viewUrl(view, { context, customViewId, defaultUrl, groupSlug })
  } else if (groupSlug) {
    return groupUrl(groupSlug)
  } else if (context === ALL_GROUPS_CONTEXT_SLUG) {
    return allGroupsUrl()
  } else if (context === PUBLIC_CONTEXT_SLUG) {
    return publicGroupsUrl()
  } else if (context === MY_CONTEXT_SLUG) {
    return myHomeUrl()
  } else {
    return defaultUrl
  }
}

export function createUrl (opts = {}, querystringParams = {}) {
  const url = baseUrl(opts) + '/create'

  return addQuerystringToPath(url, querystringParams)
}

export function createGroupUrl (opts) {
  return baseUrl(opts) + '/create/group'
}

// For specific views of a group like 'map', or 'projects'
export function viewUrl (view, { context, groupSlug, defaultUrl, customViewId }) {
  if (!view) return '/'

  const base = baseUrl({ context, groupSlug, defaultUrl })

  return `${base}/${view}${customViewId ? '/' + customViewId : ''}`
}

// Group URLS
export function groupUrl (slug, view = '', defaultUrl = allGroupsUrl()) {
  if (slug === 'public') { // TODO: remove this?
    return publicGroupsUrl()
  } else if (slug) {
    return `/groups/${slug}` + (view ? '/' + view : '')
  } else {
    return defaultUrl
  }
}

export function groupDetailUrl (slug, opts = {}, querystringParams = {}) {
  let result = baseUrl(opts)
  result = `${result}/group/${slug}`

  return addQuerystringToPath(result, querystringParams)
}

export function groupInviteUrl (group) {
  return group.invitePath ? origin() + group.invitePath : ''
}

// Post URLS
export function postUrl (id, opts = {}, querystringParams = {}) {
  const action = get('action', opts)
  let result = baseUrl(opts)
  result = `${result}/post/${id}`
  if (action) result = `${result}/${action}`

  return addQuerystringToPath(result, querystringParams)
}

export function createPostUrl (opts = {}, querystringParams = {}) {
  const url = baseUrl(opts) + '/create/post'
  return addQuerystringToPath(url, querystringParams)
}

export function editPostUrl (id, opts = {}, querystringParams = {}) {
  return postUrl(id, { ...opts, action: 'edit' }, querystringParams)
}

export function duplicatePostUrl (id, opts = {}) {
  return createPostUrl(opts, { fromPostId: id })
}

// Given a post return the the main way to view the post
// Chats go to the chat room scrolled to the post
// Posts go to the stream with the post opened
export function primaryPostUrl (post, opts = {}, querystringParams = {}) {
  let result = baseUrl(opts)
  const postId = get('id', post) || post
  if (post.type === 'chat') {
    // If topicName passed in we are opening the post from a specific chat room, otherwise if a chat then always open its first topic chat room
    const topicName = post.topics[0].name
    if (opts.commentId) {
      // When there is a commentId we pass the postId in the route params so that the post is opened when the room is loaded
      result = `${result}/chat/${topicName}/post/${postId}?commentId=${opts.commentId}`
    } else {
      // When there is no commentId, we pass the postId in the querystring so that the post is highlighted but not opened when the room is loaded
      result = `${result}/chat/${topicName}?postId=${postId}`
    }
  } else {
    result = `${result}/post/${postId}`
    if (opts.commentId) {
      result = `${result}?commentId=${opts.commentId}`
    }
  }
  return addQuerystringToPath(result, querystringParams)
}

// Messages URLs
export function messagesUrl () {
  return '/messages'
}

export function newMessageUrl () {
  return `${messagesUrl()}/new`
}

export function messageThreadUrl (id) {
  return `${messagesUrl()}/${id}`
}

export function messagePersonUrl (person) {
  // TODO: messageThreadId doesn't seem to be currently ever coming-in from the backend
  const { id: participantId, messageThreadId } = person

  return messageThreadId
    ? messageThreadUrl(messageThreadId)
    : newMessageUrl() + `?participants=${participantId}`
}

// Person URLs
export function currentUserSettingsUrl (view = 'edit-profile') {
  return '/my' + (view ? '/' + view : '')
}

export function personUrl (id, groupSlug) {
  if (!id) return '/'
  const base = baseUrl({ groupSlug })

  return `${base}/members/${id}`
}

// Topics URLs
export function topicsUrl (opts, defaultUrl = allGroupsUrl()) {
  return baseUrl({ ...opts, view: 'topics' }, defaultUrl)
}

export function topicUrl (topicName, opts) {
  return `${topicsUrl(opts)}/${topicName}`
}

export function chatUrl (chatName, { context, groupSlug }) {
  return `${baseUrl({ context, groupSlug })}/chat/${chatName}`
}

export function customViewUrl (customViewId, rootPath, opts) {
  return `${rootPath}/custom/${customViewId}`
}

export function widgetUrl ({ widget, rootPath, groupSlug: providedSlug, context = 'group' }) {
  if (!widget) return null

  const groupSlug = isContextGroupSlug(providedSlug) ? null : providedSlug
  let url = ''
  if (widget.url) return widget.url
  if (widget.view === 'about') {
    url = groupDetailUrl(groupSlug, { rootPath, groupSlug, context })
  } else if (widget.view) {
    url = viewUrl(widget.view, { groupSlug, context: widget.context || context })
  } else if (widget.viewGroup) {
    url = groupUrl(widget.viewGroup.slug)
  } else if (widget.viewUser) {
    url = personUrl(widget.viewUser.id, groupSlug)
  } else if (widget.viewPost) {
    url = postUrl(widget.viewPost.id, { groupSlug, context })
  } else if (widget.viewChat) {
    url = chatUrl(widget.viewChat.name, { rootPath, groupSlug, context })
  } else if (widget.customView) {
    url = customViewUrl(widget.customView.id, rootPath, { groupSlug })
  }

  return url
}

// URL utility functions

export function setQuerystringParam (key, value, location) {
  const querystringParams = new URLSearchParams(location.search)
  querystringParams.set(key, value)
  return querystringParams.toString()
}

export function addQuerystringToPath (path, querystringParams) {
  // The weird query needed to ignore empty arrays but allow for boolean values and numbers
  querystringParams = omitBy(x => isEmpty(x) && x !== true && !isNumber(x), querystringParams)
  return `${path}${!isEmpty(querystringParams) ? '?' + qs.stringify(querystringParams) : ''}`
}

export function removePostFromUrl (url) {
  const matchForReplaceRegex = `/post/${POST_ID_MATCH}(/.*)?`
  return url.replace(new RegExp(matchForReplaceRegex), '')
}

export function removeGroupFromUrl (url) {
  const matchForReplaceRegex = `/group/${GROUP_SLUG_MATCH}`
  return url.replace(new RegExp(matchForReplaceRegex), '')
}

export function gotoExternalUrl (url) {
  return window.open(url, null, 'noopener,noreferrer')
}

export const origin = () =>
  typeof window !== 'undefined' ? window.location.origin : host

// Utility path functions

export function isPublicPath (path) {
  return (path.startsWith('/public'))
}

export function isMapView (path) {
  return (path.includes('/map/'))
}

export function isGroupsView (path) {
  return (path.includes('/groups/'))
}
