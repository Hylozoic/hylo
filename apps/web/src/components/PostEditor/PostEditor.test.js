/* eslint-env jest */
import React, { act } from 'react'
import orm from 'store/models'
import mockGraphqlServer from 'util/testing/mockGraphqlServer'
import { graphql, HttpResponse } from 'msw'
import { render, screen, fireEvent, waitFor, AllTheProviders } from 'util/testing/reactTestingLibraryExtended'
import PostEditor from './PostEditor'
import ActionsBar from './ActionsBar'
import { editPostUrl } from '@hylo/navigation'
import * as reactRedux from 'react-redux'
import getGroupForSlug from 'store/selectors/getGroupForSlug'
import presentPost from 'store/presenters/presentPost'

jest.mock('store/actions/createPost', () => {
  return jest.fn(() => {
    return {
      type: 'CREATE_POST_SUCCESS',
      payload: { /* mock payload data */ }
    }
  })
})
import createPost from 'store/actions/createPost'

jest.mock('store/actions/updatePost', () => {
  return jest.fn(() => {
    return {
      type: 'UPDATE_POST_SUCCESS',
      payload: { /* mock payload data */ }
    }
  })
})
import updatePost from 'store/actions/updatePost'

jest.mock('store/selectors/getGroupForSlug', () => {
  return jest.fn().mockReturnValue({ id: '1', name: 'Test Group', slug: 'test-group' })
})

jest.mock('store/presenters/presentPost', () => {
  return jest.fn(() => {
    return {
      id: '1',
      type: 'request',
      title: 'Test Post'
    }
  })
})

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn().mockReturnValue({ context: 'groups', groupSlug: 'test-group' }),
  useLocation: jest.fn().mockReturnValue({ pathname: '/groups/test-group', search: '' })
}))

jest.mock('lodash/debounce', () => fn => {
  fn.cancel = jest.fn()
  return fn
})

const testGroup = {
  id: '1',
  name: 'Test Group',
  slug: 'test-group'
}

function testProviders () {
  const ormSession = orm.mutableSession(orm.getEmptyState())
  ormSession.Me.create({
    id: '20',
    memberships: [ormSession.Membership.create({
      id: '345',
      group: '1'
    })]
  })
  ormSession.Group.create(testGroup)
  ormSession.Post.create({
    id: '1',
    title: 'Test Post',
    type: 'discussion',
    groups: [{ id: '1', name: 'Test Group' }],
    topics: [{ name: 'design' }],
    creator: { id: '20' }
  })
  const reduxState = { orm: ormSession.state }
  return AllTheProviders(reduxState)
}

beforeEach(() => {
  mockGraphqlServer.use(
    // Log all GraphQL operations
    graphql.operation(({ query, variables }) => {
      // console.log('GraphQL Operation:', { query, variables })
      return HttpResponse.json({ data: {} })
    }),
    graphql.query('FetchPost', ({ query, variables }) => {
      return HttpResponse.json({
        data: {
          post: {
            id: '1',
            title: 'Test Post',
            type: 'discussion',
            groups: [{ id: '1', name: 'Test Group' }],
            topics: [{ name: 'design' }]
          }
        }
      })
    }),
    graphql.query('FetchTopics', ({ query, variables }) => {
      return HttpResponse.json({
        data: {
          topics: []
        }
      })
    }),
    graphql.mutation('CreatePost', ({ query, variables }) => {
      return HttpResponse.json({
        data: {
          createPost: {
            id: '1',
            title: 'Test Post',
            groups: [{ id: '1', name: 'Test Group' }],
            topics: [{ name: 'design' }]
          }
        }
      })
    })
  )
})

const renderComponent = (props = {}) => {
  return render(
    <PostEditor {...props} />,
    { wrapper: testProviders() }
  )
}

describe('PostEditor', () => {

  it('renders with min props', async () => {
    renderComponent()

    await waitFor(() => {
      expect(screen.getByLabelText(/title/i, { selector: 'input' })).toBeInTheDocument()
    })
    expect(screen.getByLabelText(/title/i, { })).toBeInTheDocument()
  })

  test.each(['discussion', 'request', 'offer', 'resource'])(
    'renders correct title input for post type %s',
    async (type) => {
      renderComponent({ post: { type } })
      await waitFor(() => {
        expect(screen.getByLabelText(/title/i, { })).toBeInTheDocument()
      })
    }
  )

  it('calls createPost when saving a new post', async () => {
    jest.spyOn(require('react-router-dom'), 'useParams').mockReturnValue({ groupSlug: 'test-group' })

    renderComponent()

    const titleInput = await screen.findByLabelText(/title/i, { })
    fireEvent.change(titleInput, { target: { value: 'Test Title' } })
    fireEvent.click(screen.getByLabelText(/submit/i))

    await waitFor(() => {
      expect(createPost).toHaveBeenCalled()
    })
  })

  describe('for a new event', () => {
    it('renders correctly', async () => {
      renderComponent({ post: { type: 'event', groups: [] } })

      await waitFor(() => {
        expect(screen.getByText('Timeframe')).toBeInTheDocument()
        expect(screen.getByText('Location')).toBeInTheDocument()
      })
    })
  })

  describe('editing a post', () => {
    const editProps = {
      editing: true,
      post: {
        id: '1',
        type: 'request',
        title: 'Test Post'
      },
      setIsDirty: jest.fn()
    }

    it('calls updatePost when saving an edited post', async () => {
      jest.clearAllMocks()
      jest.spyOn(require('react-router-dom'), 'useParams').mockReturnValue({ groupSlug: 'test-group', postId: '1' })
      renderComponent(editProps)

      const titleInput = await screen.findByLabelText(/title/i)
      expect(titleInput.value).toEqual('Test Post')
      fireEvent.change(titleInput, { target: { value: 'New Title' } })
      expect(titleInput.value).toEqual('New Title')

      const submitButton = await screen.findByRole('button', { name: /submit/i })
      expect(submitButton).toBeEnabled()
      fireEvent.click(submitButton)

      await waitFor(() => {
        expect(createPost).not.toHaveBeenCalled()
        expect(updatePost).toHaveBeenCalled()
      })
    })

    it('tracks dirty state when content changes', async () => {
      renderComponent(editProps)
      await waitFor(() => {
        expect(screen.getByDisplayValue('Test Post')).toBeInTheDocument()
      })
      fireEvent.change(screen.getByDisplayValue('Test Post'), { target: { value: 'New Title' } })
      await waitFor(() => {
        expect(editProps.setIsDirty).toHaveBeenCalled()
      })
    })

    it('shows error for invalid title length', async () => {
      renderComponent(editProps)
      await waitFor(() => {
        expect(screen.getByDisplayValue('Test Post')).toBeInTheDocument()
      })
      await waitFor(() => {
        fireEvent.change(screen.getByDisplayValue('Test Post'), { target: { value: 'x'.repeat(81) } })
      })
      await waitFor(() => {
        expect(screen.getByText('Title limited to 80 characters')).toBeInTheDocument()
      })
    })
  })
})

describe('ActionsBar', () => {
  const baseProps = {
    id: '1',
    addAttachment: jest.fn(),
    showImages: false,
    showFiles: false,
    valid: true,
    loading: false,
    submitButtonLabel: 'Post',
    save: jest.fn(),
    setAnnouncementSelected: jest.fn(),
    announcementSelected: false,
    toggleAnnouncementModal: jest.fn(),
    showAnnouncementModal: false,
    groupCount: 1,
    canMakeAnnouncement: true,
    myAdminGroups: [],
    groups: [],
    invalidPostWarning: 'Invalid post',
    t: jest.fn(str => str)
  }

  it('renders correctly', async () => {
    render(<ActionsBar {...baseProps} />)
    await waitFor(() => {
      expect(screen.getByRole('button')).toBeInTheDocument()
      expect(screen.getByTestId('add-image-icon')).toBeInTheDocument()
      expect(screen.getByTestId('add-file-icon')).toBeInTheDocument()
    })
  })

  it('disables post button when invalid', async () => {
    render(<ActionsBar {...baseProps} valid={false} />)
    await waitFor(() => {
      expect(screen.getByRole('button')).toHaveClass('disabled')
    })
  })

  it('shows announcement icon when user can make announcements', async () => {
    render(<ActionsBar {...baseProps} canMakeAnnouncement />)
    await waitFor(() => {
      expect(screen.getByTestId('announcement-icon')).toBeInTheDocument()
    })
  })

  it('does not show announcement icon when user cannot make announcements', async () => {
    render(<ActionsBar {...baseProps} canMakeAnnouncement={false} />)
    await waitFor(() => {
      expect(screen.queryByTestId('announcement-icon')).not.toBeInTheDocument()
    })
  })
})
