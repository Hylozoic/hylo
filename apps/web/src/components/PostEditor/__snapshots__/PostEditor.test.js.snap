// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostEditor for a new event renders correctly 1`] = `
<div
  data-stylename="wrapper"
>
  <div
    data-stylename="header"
  >
    <div
      data-stylename="initial"
    >
      <div />
    </div>
  </div>
  <div
    data-stylename="body"
  >
    <div
      data-stylename="body-column"
    >
      <RoundImage
        data-stylename="titleAvatar"
        medium={true}
      />
    </div>
    <div
      data-stylename="body-column"
    >
      <input
        data-stylename="titleInput"
        disabled={false}
        maxLength={80}
        onChange={[Function]}
        placeholder="Add a title"
        type="text"
        value=""
      />
      <ForwardRef(HyloEditor)
        data-stylename="editor"
        onAddLink={[Function]}
        onAddTopic={[Function]}
        onUpdate={[Function]}
        placeholder="Add a description"
        readOnly={false}
        showMenu={true}
      />
    </div>
  </div>
  <Connect(AttachmentManager)
    attachmentType="image"
    showAddButton={true}
    showLabel={true}
    showLoading={true}
    type="post"
  />
  <Connect(AttachmentManager)
    attachmentType="file"
    showAddButton={true}
    showLabel={true}
    showLoading={true}
    type="post"
  />
  <div
    data-stylename="footer"
  >
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Topics
      </div>
      <div
        data-stylename="footerSection-topics"
      >
        <Connect(TopicSelector)
          forGroups={Array []}
          onChange={[Function]}
        />
      </div>
    </div>
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Post in
        *
      </div>
      <div
        data-stylename="footerSection-groups"
      >
        <GroupsSelector
          onChange={[Function]}
          options={Array []}
          readOnly={false}
          selected={Array []}
          t={[Function]}
        />
      </div>
    </div>
    <PublicToggle
      isPublic={false}
      t={[Function]}
      togglePublic={[Function]}
    />
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Timeframe
      </div>
      <div
        data-stylename="datePickerModule"
      >
        <DatePicker
          onChange={[Function]}
          placeholder="Select Start"
          value={"2024-07-23T16:30:00.000Z"}
        />
        <div
          data-stylename="footerSection-helper"
        >
          To
        </div>
        <DatePicker
          onChange={[Function]}
          placeholder="Select End"
          value={"2024-07-23T16:30:00.000Z"}
        />
      </div>
    </div>
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Invite People
      </div>
      <div
        data-stylename="footerSection-groups"
      >
        <Connect(MemberSelector)
          forGroups={Array []}
          initialMembers={Array []}
          onChange={[Function]}
          readOnly={false}
        />
      </div>
    </div>
    <ActionsBar
      canMakeAnnouncement={true}
      groupCount={0}
      groups={Array []}
      invalidPostWarning="You need a title and at least one group to post"
      loading={false}
      save={[Function]}
      submitButtonLabel="Post"
      t={[Function]}
      toggleAnnouncementModal={[Function]}
      valid={false}
    />
  </div>
</div>
`;

exports[`PostEditor renders announcement option with admin in props 1`] = `
<div
  data-stylename="actionsBar"
>
  <div
    data-stylename="actions"
  >
    <Connect(UploadAttachmentButton)
      allowMultiple={true}
      attachmentType="image"
      id={1}
      onSuccess={[Function]}
      type="post"
    >
      <Icon
        data-stylename="action-icon"
        name="AddImage"
      />
    </Connect(UploadAttachmentButton)>
    <Connect(UploadAttachmentButton)
      allowMultiple={true}
      attachmentType="file"
      id={1}
      onSuccess={[Function]}
      type="post"
    >
      <Icon
        data-stylename="action-icon"
        name="Paperclip"
      />
    </Connect(UploadAttachmentButton)>
    <span
      data-tooltip-id="announcement-tt"
      data-tooltip-content="Send Announcement"
    >
      <Icon
        data-stylename="action-icon"
        name="Announcement"
        onClick={[Function]}
      />
      <ReactTooltip
        clickable={false}
        delayShow={550}
        effect="solid"
        id="announcement-tt"
        insecure={true}
        resizeHide={true}
        wrapper="div"
      />
    </span>
  </div>
  <Button
    color="green"
    data-stylename="postButton"
    dataFor="submit-tt"
    dataTip=""
    disabled={false}
    label="Save"
    onClick={[Function]}
  />
  <Tooltip
    delay={150}
    id="submit-tt"
    offset={
      Object {
        "top": 0,
      }
    }
    position="bottom"
  />
</div>
`;

exports[`PostEditor renders with min props 1`] = `
<div
  data-stylename="wrapper"
>
  <div
    data-stylename="header"
  >
    <div
      data-stylename="initial"
    >
      <div>
        <Button
          borderRadius="5px"
          className=""
          color=""
          disabled={false}
          key="discussion"
          label={
            <span
              data-stylename="initial-prompt"
            >
              <span>
                discussion
              </span>

              <Icon
                data-stylename="icon icon-discussion"
                name="ArrowDown"
              />
            </span>
          }
          noDefaultStyles={true}
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    data-stylename="body"
  >
    <div
      data-stylename="body-column"
    >
      <RoundImage
        data-stylename="titleAvatar"
        medium={true}
      />
    </div>
    <div
      data-stylename="body-column"
    >
      <input
        data-stylename="titleInput"
        disabled={false}
        maxLength={80}
        onChange={[Function]}
        placeholder="Add a title"
        type="text"
        value=""
      />
      <ForwardRef(HyloEditor)
        contentHTML=""
        data-stylename="editor"
        onAddLink={[Function]}
        onAddTopic={[Function]}
        onUpdate={[Function]}
        placeholder="Add a description"
        readOnly={false}
        showMenu={true}
      />
    </div>
  </div>
  <Connect(AttachmentManager)
    attachmentType="image"
    showAddButton={true}
    showLabel={true}
    showLoading={true}
    type="post"
  />
  <Connect(AttachmentManager)
    attachmentType="file"
    showAddButton={true}
    showLabel={true}
    showLoading={true}
    type="post"
  />
  <div
    data-stylename="footer"
  >
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Topics
      </div>
      <div
        data-stylename="footerSection-topics"
      >
        <Connect(TopicSelector)
          forGroups={Array []}
          onChange={[Function]}
        />
      </div>
    </div>
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label"
      >
        Post in
        *
      </div>
      <div
        data-stylename="footerSection-groups"
      >
        <GroupsSelector
          onChange={[Function]}
          options={Array []}
          readOnly={false}
          selected={Array []}
          t={[Function]}
        />
      </div>
    </div>
    <PublicToggle
      isPublic={false}
      t={[Function]}
      togglePublic={[Function]}
    />
    <div
      data-stylename="footerSection"
    >
      <div
        data-stylename="footerSection-label alignedLabel"
      >
        Location
      </div>
      <Connect(LocationInput)
        onChange={[Function]}
        placeholder="Where is your {{type}} located?"
        saveLocationToDB={true}
      />
    </div>
    <ActionsBar
      canMakeAnnouncement={true}
      groupCount={0}
      groups={Array []}
      invalidPostWarning="You need a title and at least one group to post"
      loading={false}
      save={[Function]}
      submitButtonLabel="Post"
      t={[Function]}
      toggleAnnouncementModal={[Function]}
      valid={false}
    />
  </div>
</div>
`;
